src/reducers/tree.ts

```diff
@@ -0,0 +1,178 @@
+import { AnyAction } from "@reduxjs/toolkit";
+import { NODE_NOT_VISIBLE, NODE_VISIBLE, NODE_VISIBLE_TO_MAP_ONLY } from "../util/globals";
+import { countTraitsAcrossTree } from "../util/treeCountingHelpers";
+import { addNodeAttrs } from "../util/treeMiscHelpers";
+import * as types from "../actions/types";
+import { PhyloNode } from "../components/tree/phyloTree/types";
+
+/**
+ * Maps mutation strings (in format gene:fromPosTo, e.g. 'nuc:A123T')
+ * to their occurrence count in the tree
+ */
+export type Mutations = Record<string, number>
+
+export interface ReduxNode {
+  /** the index of the node in the nodes array. set so that we can access visibility / nodeColors if needed */
+  arrayIdx?: number
+
+  branch_attrs?: {
+    mutations?: {
+      [gene: string]: string[]
+    }
+    labels?: Record<string, any>
+  }
```

Comment:
[not a request for changes]

This reminded me of Ivan's advice to type the underlying Auspice JSON and let types flow through from there. After doing this work do you think we've essentially typed the Auspice JSON by typing the structure it's processed into?

Also, `processBranchLabelsInPlace` turns this into `labels?: Record<string, string>` so something to consider when we start to distinguish between types in the data-constructing phase vs data-consuming phase.

Reply:
> For example, the bulk of [Metadata](https://github.com/nextstrain/auspice/blob/d5986de2d645958c68fc3f59fc3e1dd967eecb5e/src/metadata.ts#L3-L8) is still missing

I [started down this road a while ago](https://github.com/nextstrain/auspice/tree/james/ts), largely following Ivan's advice to start with the JSON and let types flow down. We/I should revisit!

Reply:
Nice, I wasn't aware of that adventure! These look familiar... https://github.com/nextstrain/auspice/blob/ce33cbd65975e3891bb418b7fe6f3496db77e5cc/src/reducers/metadata.ts#L12-L31

Reply:
Yeah it was a bit of an adventure. Be fun to revisit it with the knowledge you've learned. I hadn't looked at it in months. This comment stood out](https://github.com/nextstrain/auspice/compare/master...james/ts#diff-c39c21eb00b72bc5c073924636b033832877770bc0b166a23654a99924be368fR7-R9) upon re-reading:

> Does it make sense to have two state types, a non-loaded one and a loaded one (discriminated type based on loaded: boolean) to give more assurances? Otherwise most properties are potential undefineds.